### DISCUSSION TITLE
Cockroach DB Slow Writes/Inserts Compared To MariaDB & Timescale DB

### DISCUSSION THREAD

**Discussion by Zub4ir** [cockroachdb/cockroach#94989] [Q&A]
Hi,

I work in an investment house where we process time series data and fund position data for quantitative analytics and data science. We query about 1800 time series (4.5mil rows), run some calculations and insert the results multiple times per hour. About 200 of these processes run throughout the day. We also have on-the-fly reports/analysis/optimizations generated on our analytics platform.

I replicated a simple business use case as a benchmark for MariaDB, CRDB and Timescale DB. I setup pretty much standard tables for all three - no fancy indexes or hyper tables from timescale. I used two types of data sets

- time series data - long narrow tables, star schema, 1 fact table for long time series data, and 3 dim tables
- wide tables data - long wide tables, not efficiently stored, various data types, 2 tables

Both Maria DB and Timescale is "single node". CRDB is running as 3 nodes.

Can anyone explain the differences here, specifically as to why CRDB is so slow with inserts?

All inserts/queries were run in Python3 using `sqlalchemy create_engine` , `pandas read_sql_query` and `DataFrame.to_sql`. Below is the elapsed time and rows per second (r/s) for each benchmark. Less time elapsed and more r/s is better.

![image](https://user-images.githubusercontent.com/36241875/211532079-eccf94ea-d1a3-4427-83e0-7918bd46654f.png)

![image](https://user-images.githubusercontent.com/36241875/211529250-39777a15-2b46-4f4e-83b3-c5f57494c29c.png)

I know this is probably lacking in detail which I can share. But understand that these are simple setups and using out-of-the-box configs. I do not want to jump through hoops to optimize given that other products are out-of-the-box faster with simpler and more convenient setups.

**Comment by zjyuan** [comment]
I think it's because crdb is a distributed newsql database. Your data in crdb is stored in a sorted map. This sorted map is divided into ranges. Each range is replicated 3 times by default. So for each write, there is overhead of replication compared to monolithic database.

**Comment by Zub4ir** [comment]
I ran it on a single node setup as well. Performance was slightly worse than 3 node.

**Reply by zjyuan** [reply]
I am a learner of crdb and would like to help. Would you mind to share your data and setup?

**Comment by gnat** [comment]
Cockroach is going to need more hardware to scale your writes- Your replication factor determines your "quorum group" size: you need more "quorum groups".

As of the current day, v23, those numbers are fairly decent for one "quorum group".

You get stronger `INSERT` performance by:

* Only creating indexes you absolutely need because each index is +1 `INSERT` (each index is another small table.. this can feel like a performance cliff if you have many).
* Leverage `gen_random_uuid()` as it can write right away and bypass the  "quorum group" consistency checks.

I wish CockroachDB had a way to trade consistency for `INSERT` speed when desired, like ScyllaDB, or have more types that bypassed the quorum check like ULID. Any patterns to help increase performance are much welcomed.

**Comment by Yash-xoxo** [comment]
The performance gap you're seeing is expected for a few reasons:

**Why CRDB is slower for inserts:**

1. **Distributed overhead** - Even single-node CRDB still does:
   - Raft consensus (3x replication by default)
   - Range splitting and balancing
   - Transaction coordination

2. **Time series anti-pattern** - As @zjyxn mentioned, sorted sequential inserts (by timestamp) create hot spots on a single range, causing write contention

3. **Default replication factor** - CRDB replicates 3x even in single-node mode

**To improve CRDB performance:**
```sql
-- Reduce replication for single node
ALTER RANGE default CONFIGURE ZONE USING num_replicas = 1;

-- Add index only if absolutely needed (each index = write amplification)
-- Consider hash-sharded index to distribute writes
CREATE INDEX ON timeseries (timestamp) USING HASH WITH BUCKET_COUNT = 8;

-- Use IMPORT for bulk data
IMPORT INTO timeseries CSV DATA ('data.csv');
```

**Better patterns for time series:**

- Use `gen_random_uuid()` or hash-based primary keys instead of sequential timestamps
- Leverage `AS OF SYSTEM TIME` for time-travel queries without indexes
- Batch inserts where possible

**Reality check:** 
MariaDB and TimescaleDB are optimized for single-node time series workloads. CRDB trades insert speed for:
- Distributed scale-out
- Strong consistency across regions
- Automatic rebalancing

If you don't need CRDB's distributed features, stick with TimescaleDB for time series. If you do, the performance trade-off is the cost of distributed guarantees.
